code/datums/components/nanites.dm

```diff
@@ -0,0 +1,216 @@
+	var/regen_rate = 0.5		//nanites generated per second
+	var/safety_threshold = 50	//how low nanites will get before they stop processing/triggering
+	var/cloud_id = 0 			//0 if not connected to the cloud, 1-100 to set a determined cloud backup to draw from
+	var/next_sync = 0
+	var/list/datum/nanite_program/programs = list()
+	var/max_programs = 15
+
+	var/stealth = FALSE //if TRUE, does not appear on HUDs and health scans, and does not display the program list on nanite scans
+
+/datum/component/nanites/Initialize(amount = 100, cloud = 0)
+	nanite_volume = amount
+	cloud_id = cloud
+
+	RegisterSignal(parent, COMSIG_NANITE_GET_DATA, .proc/get_data)
+	RegisterSignal(parent, COMSIG_NANITE_GET_PROGRAMS, .proc/get_programs)
+	RegisterSignal(parent, COMSIG_NANITE_SET_VOLUME, .proc/set_volume)
+	RegisterSignal(parent, COMSIG_NANITE_ADJUST_VOLUME, .proc/adjust_nanites)
+	RegisterSignal(parent, COMSIG_NANITE_SET_MAX_VOLUME, .proc/set_max_volume)
+	RegisterSignal(parent, COMSIG_NANITE_SET_CLOUD, .proc/set_cloud)
+	RegisterSignal(parent, COMSIG_NANITE_SET_SAFETY, .proc/set_safety)
+	RegisterSignal(parent, COMSIG_NANITE_SET_REGEN, .proc/set_regen)
+	RegisterSignal(parent, COMSIG_NANITE_ADD_PROGRAM, .proc/add_program)
+	RegisterSignal(parent, COMSIG_NANITE_SYNC, .proc/sync)
+
+	//Nanites without hosts are non-interactive through normal means
+	if(isliving(parent))
+		host_mob = parent
+
+		if(!(MOB_ORGANIC in host_mob.mob_biotypes) && !(MOB_UNDEAD in host_mob.mob_biotypes)) //Shouldn't happen, but this avoids HUD runtimes in case a silicon gets them somehow.
+			return COMPONENT_INCOMPATIBLE
+
+		host_mob.hud_set_nanite_indicator()
+		START_PROCESSING(SSnanites, src)
+		RegisterSignal(host_mob, COMSIG_ATOM_EMP_ACT, .proc/on_emp)
+		RegisterSignal(host_mob, COMSIG_MOB_DEATH, .proc/on_death)
+		RegisterSignal(host_mob, COMSIG_MOB_ALLOWED, .proc/check_access)
+		RegisterSignal(host_mob, COMSIG_LIVING_ELECTROCUTE_ACT, .proc/on_shock)
+		RegisterSignal(host_mob, COMSIG_LIVING_MINOR_SHOCK, .proc/on_minor_shock)
+		RegisterSignal(host_mob, COMSIG_MOVABLE_HEAR, .proc/on_hear)
+		RegisterSignal(host_mob, COMSIG_SPECIES_GAIN, .proc/check_viable_biotype)
+
+		RegisterSignal(host_mob, COMSIG_NANITE_SIGNAL, .proc/receive_signal)
+		if(cloud_id)
+			cloud_sync()
+	else if(!istype(parent, /datum/nanite_cloud_backup))
+		return COMPONENT_INCOMPATIBLE
+
+/datum/component/nanites/Destroy()
+	STOP_PROCESSING(SSnanites, src)
+	QDEL_LIST(programs)
+	if(host_mob)
+		host_mob.hud_set_nanite_indicator()
+	host_mob = null
+	return ..()
+
+/datum/component/nanites/InheritComponent(datum/component/nanites/new_nanites, i_am_original, list/arguments)
+	if(new_nanites)
+		adjust_nanites(new_nanites.nanite_volume)
+	else
+		adjust_nanites(arguments[1]) //just add to the nanite volume
+
+/datum/component/nanites/process()
+	adjust_nanites(regen_rate)
+	for(var/X in programs)
+		var/datum/nanite_program/NP = X
+		NP.on_process()
+	host_mob.diag_hud_set_nanite_bar()

```

Comment:
Unless you found some other maintainer to say otherwise this needs to be in the component